### DISCUSSION TITLE
Something wrong with the unit of weight

### DISCUSSION THREAD

**Discussion by HaoPengYY** [SmileiPIC/Smilei#783] [Q&A]
Dear Developers and Users,

Hi, I was trying to dump the trajectories of a bunch of test electrons in a 3D laser-plasma interaction simulation and use the information including the particles weights for post-processing. But when I used h5py to open the trajectory file and get out the data, I found that the unit of the weight is different from what I calculated, while other data including the positions and fields agreed well with my calculation. It seems there is something wrong with the out weight unit in SMILEI.

Attached is the input file of the simulation I run. The normalized weight of 'teon' output from Happi is about 0.1211, which agrees well wight the calculation, but the unit stored in the hdf5 trajectory file,  i.e. the attributes 'unitSI' of the weight dataset is wrong ,which is about the 25 orders of magnitude above the value I calculate according to the formula below:
<img width="537" alt="image" src="https://github.com/user-attachments/assets/8a926c09-f311-4fe4-8fd8-8587e010d0f5" />
[Uploading input.py.zip…]()

**Comment by HaoPengYY** [comment]
It seems I did not successfully uploaded the input file, re-upload here..
[input.py.zip](https://github.com/user-attachments/files/18271827/input.py.zip)

**Comment by mccoys** [comment]
Please do not consider the 'unitSI' parameter. It is only a dummy value for conforming with the openPMD standard.

**Reply by HaoPengYY** [reply]
But, it is weird, right? For all the other parameters, 'x', 'y', 'z', 'E/x', E/y', E/z', B/x', B/y', B/z', and momenta, 'unitSI' agrees well with my expectation (when the reference frequency is provided).  I suppose there will be something wrong with the unit of weight dumped with setting "Pint=True"?

**Reply by mccoys** [reply]
The value of unitSI in the hdf5 file is calculated here and it seems fine to me: https://github.com/SmileiPIC/Smilei/blob/3989d925ea35616c3068e4ce176bd2444eb98fa1/src/Params/OpenPMDparams.cpp#L80

I am not sure I get your point exactly. You say this is the value in the hdf5 file thus unrelated to happi, but then you suggest it could be caused by `pint=True` which is happi-only. Are you checking in the hdf5 file directly without using happi?

**Reply by HaoPengYY** [reply]
In the code you shown above,  `5.034163 * Wr*Wr` should be the density unity. But it is weird, in the same code block you defined the density unit to be:
```
        } else if( unit_type == SMILEI_UNIT_DENSITY ) {
            unitDimension[unit_type] = { -3., 0., 1., 1., 0., 0., 0. };
            unitSI[unit_type] = 5.034163 * Wr*Wr; // e0 * me * Wr^2 / e
```
The density unit should be `e0 * me * Wr^2 / e^2`, not `e0 * me * Wr^2 / e`,  right? You forgot to put a power to `e`? Moreover, `e0 * me / e` equals to 5.034163e-23, not 5.034163.

**Reply by mccoys** [reply]
Ok you are probably right. I think there is a confusion between charge density and number density. But even then the 1e23 factor is missing. Let me add this as an issue
